### ISSUE TITLE
TCH003 false positive when typing nested functions

### ISSUE THREAD

**Issue by VoodaGod** [charliermarsh/ruff#4020]
consider `test.py`
```
from collections.abc import Callable, Collection


def test():
    def nested() -> Callable[[Collection[str]], list[str]]:
        def double_nested(x: Collection[str]) -> list[str]:
            return list(x)

        return double_nested

    a = nested()
    return a(("a", "b"))


print(test())

```
outputs
```
['a', 'b']
```

ruff complains:
```
test.py:1:29: TCH003 Move standard library import `collections.abc.Callable` into a type-checking block  
test.py:1:39: TCH003 Move standard library import `collections.abc.Collection` into a type-checking block
```

appeasing ruff
```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable, Collection


def test():
    def nested() -> Callable[[Collection[str]], list[str]]:
        def double_nested(x: Collection[str]) -> list[str]:
            return list(x)

        return double_nested

    a = nested()
    return a(("a", "b"))


print(test())

```
breaks the code:
```
Traceback (most recent call last):
  File "c:\Programs\PCGUserdir\Repos\SmartSensorBus_configurator\tempCodeRunnerFile.py", line 18, in <module>
    print(test())
          ^^^^^^
  File "c:\Programs\PCGUserdir\Repos\SmartSensorBus_configurator\tempCodeRunnerFile.py", line 8, in test
    def nested() -> Callable[[Collection[str]], list[str]]:
                    ^^^^^^^^
NameError: name 'Callable' is not defined. Did you mean: 'callable'?
```

the command i used to invoke ruff: `python -m ruff .`
my pyproject.toml ruff settings:
```
[tool.ruff]
line-length = 100
ignore-init-module-imports = true
extend-select = [
	"W",      # pydocstyle warnings
	"C90",    # mccabe complexity
	"I",      # import sorting
	"N",      # pep8 naming
	"UP",     # old syntax/constructs
	"YTT",    # old syntax/constructs
	"B",      # common sources of bugs
	"A",      # shadowing builtins
	"COM",    # trailing commas
	"C4",     # comprehensions
	"EXE",    # shebangs / executable .py files
	"ISC",    # implicit string concatenation
	"PT",     # pytest
	"Q",      # quotes
	"RET502", # missing return
	"RET503", # missing return
	"SIM",    # simplify code
	"TCH",    # type checking imports
	"ARG",    # unused arguments
	"PTH",    # use pathlib
	"PLE",    # pylint errors
]
ignore = [
	"E501",   # let black handle line length
	"COM812", # let black handle missing trailing commas
]
unfixable = [
	"F841", # unused variable
	"F401", # unused import
]
```

ruff version: `ruff 0.0.261`
python version: `Python 3.11.2`

**Comment by charliermarsh** [comment]
Oh interesting, I thought annotations for functions within functions weren't required to be available at runtime, but I must've misunderstood. Thanks!
